Add ERC-2212 markdown (#2444)
ethereum/EIPs: EIPS/eip-2212.md

---
eip: 2212
title: ERC-2212 Interest Earning Stakes
author: Paul Razvan Berg (@PaulRBerg) <paul@sablier.finance>
discussions-to: https://github.com/ethereum/EIPs/issues/2212
status: Draft
type: Standards Track
category: ERC
created: 2019-07-26
requires: 20
---

## Simple Summary

We propose a new way to monetise decentralised apps that involves users staking tokens in exchange for a product or a service and creators earning interest on the pooled stakes through a continuous lending protocol such as [Compound](https://compound.finance).

## Abstract

> The advent of decentralised lending brought to the web3 ecosystem a new primitive for designing business models. We herein describe a smart contract interface that accepts deposits ("stakes") in ERC20 tokens. These deposits are immediately lent on the market to earn interest either for just the owner of the contract, or for both them and the user. We assume that users get access to a product or a service for as long as they keep staking tokens.

Reviewer: ```suggestion
The advent of decentralised lending brought to the web3 ecosystem a new primitive for designing business models. We herein describe a smart contract interface that accepts deposits ("stakes") in [ERC-20](https://eips.ethereum.org/EIPS/eip-20) tokens. These deposits are immediately lent on the market to earn interest either for just the owner of the contract, or for both them and the user. We assume that users get access to a product or a service for as long as they keep staking tokens.
```

### Methods

#### owner

Returns the owner of the contract.

```solidity
function owner() external view returns (address);
```

#### fee

Returns the fee as a percentage value scaled by 1e18.

```solidity
function fee() external view returns (uint256);
```

#### balanceOf

Returns the stake that consists of the initial deposit plus the interest accrued over time.

SHOULD revert if `staker` doesn't have an active stake. SHOULD revert if `tokenAddress` doesn't match either the cToken or the underlying of the stake.

```solidity
function balanceOf(address staker, address tokenAddress) external view returns (uint256);
```

#### depositOf

Returns the initial deposit.

SHOULD revert if `staker` doesn't have an active stake. SHOULD revert if `tokenAddress` doesn't match either the cToken or the underlying of the stake.

```solidity
function depositOf(address staker, address tokenAddress) external view returns (uint256);
```

#### whitelistCToken

> Whitelists a cToken to automatically convert its underlying when deposited.

Reviewer: What is a cToken?

```

#### DiscardCToken

MUST be triggered when `discardCToken` is successfully called.

```solidity
event DiscardCToken(address indexed cTokenAddress);
```

#### Stake

MUST be triggered when `stake` is successfully called.

```solidity
event Stake(address indexed staker, address indexed cTokenAddress, address indexed underlyingAddress, bool converted, uint256 cTokenDeposit, uint256 exchangeRate);
```

#### Redeem

MUST be triggered when `redeem` is successfully called.

```solidity
event Redeem(address indexed staker, address indexed cTokenAddress, address indexed underlyingAddress, uint256 cTokenFee, uint256 cTokenWithdrawal, uint256 exchangeRate);
```

#### TakeEarnings

MUST be triggered when `takeEarnings` is successfully called.

```solidity
event TakeEarnings(address indexed tokenAddress, uint256 indexed amount);
```

#### UpdateFee

MUST be triggered when `updateFee` is successfully called.

```solidity
event UpdateFee(uint256 indexed fee);
```

## Rationale

We designed these interest earning stakes with simplicity in mind, but we acknowledge that there are some missing features and obvious concerns which we discuss below.

> ---

Reviewer: ```suggestion
```

It is worth it to mention PoolTogether, a "no-loss lottery" that launched recently. It works by having participants deposit money in their contracts, locking it up for a while and earning interest through Compound. Ultimately, they choose a lucky winner and return the deposits back to all other players. PoolTogether takes a 10% commission on the prize.

We wrote this specification because interest earning stakes are a powerful financial instrument and a community-vetted standard and implementation will reduce the cost of replication.

### Use Cases

- Software products and services
- Charity pools
- Games
- Gym memberships

## Specification

Assigning an owner is required for this spec to make sense, thus we assume that a proper implementation would use a
well-known contract such as OpenZeppelin's
[Ownable](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/ownership/Ownable.sol). For
brevity, we haven't included any method for assigning or transferring ownership.

> Every time we refer to cTokens, we mean Compound Tokens, as defined in the official

Reviewer: Hm, I don't see a specification there. What is the interface of a cToken and why does ERC-2212 need to use cTokens and not any ERC-20 ?

Author: Afaik, Compound did not publish the interface for cTokens in this repo, only on their website.

This spec can't use ERC-20, because of the need to calculate interest and withhold a portion of it. Then, I didn't include other protocols other than Compound merely because it would've taken more time and I wanted to get it out there and get feedback.

If you read the latest messages in https://github.com/ethereum/EIPs/issues/2212, you can  see that we started working on a v2, which is protocol-agnostic. v1 is not.

> [documentation](https://compound.finance/ctokens). These are ERC-20 tokens with an interest accrual feature built in.

Reviewer: This link is now dead.

Author: Is it? I just opened it and it worked for me.

Reviewer: It doesn't go to any kind of documentation, just a marketing page. There should be a link to a specification, if it cannot be another ERC.

Author: Fair enough.

### Methods

#### owner

Returns the owner of the contract.

```solidity
function owner() external view returns (address);
```

#### fee

Returns the fee as a percentage value scaled by 1e18.

```solidity
function fee() external view returns (uint256);
```

#### balanceOf

Returns the stake that consists of the initial deposit plus the interest accrued over time.

SHOULD revert if `staker` doesn't have an active stake. SHOULD revert if `tokenAddress` doesn't match either the cToken or the underlying of the stake.

```solidity

> function balanceOf(address staker, address tokenAddress) external view returns (uint256);

Reviewer: Since the ABI encoding ignores the parameter names, this may be something prone to collisions.

```

#### depositOf

Returns the initial deposit.

SHOULD revert if `staker` doesn't have an active stake. SHOULD revert if `tokenAddress` doesn't match either the cToken or the underlying of the stake.

```solidity
function depositOf(address staker, address tokenAddress) external view returns (uint256);
```

#### whitelistCToken

Whitelists a cToken to automatically convert its underlying when deposited.

MUST revert if `msg.sender` is not the owner.

**Triggers Event**: [WhitelistCToken](#whitelistctoken)

```solidity
function whitelistCToken(address cTokenAddress, address underlyingAddress) external;
```

#### discardCToken

Discards a cToken that has been whitelisted before.

MUST revert if `msg.sender` is not the owner.

**Triggers Event**: [DiscardCToken](#discardctoken)

```solidity

> function discardCToken(address cTokenAddress) external

Reviewer: ```suggestion
function discardCToken(address cTokenAddress) external;
```

> function discardCToken(address cTokenAddress) external

Reviewer: I think it would be better using the type system if there is a requirement to use "cToken"s.

#### Rent-Seeking

Interest earning stakes work best for products or services that are priced on time instead of quantity (imagine subscription-based SaaS). If a dapp is quota-based, its creators would be better off charging a fee on every transfer.

## Implementation

[Head to the Sablier organisation on GitHub](https://github.com/sablierhq/erc2212) for a WIP implementation.

## Additional References

- [Thread on cTokens](https://twitter.com/PaulRBerg/status/1147980107740524544)
- [Compound Protocol](https://github.com/compound-finance/compound-protocol)
- [PoolTogether](https://pooltogether.us)
- [Pooled cDAI](https://github.com/ZeframLou/pooled-cdai)
- [Lending Proxy by Luciano Bertenasco](https://github.com/lbertenasco/lending-proxy)
- [Web3 Revenue Primitives](https://github.com/FEMBusinessModelsRing/web3_revenue_primitives)

## Copyright

Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

> @PaulRBerg and @SablierHQ.

Reviewer: What are these for?

Author: I saw the copyright in other standards and wrote one here too. I'll remove if its place is not here.

Reviewer: There is no copyright with CC0, if I understand it correctly.

Author: That is why it is "waived" - it's a small thing, happy to remove it if you think it is superfluous.

Reviewer: @MicahZoltu @lightclient any comments here?
